File: LevelOrderTraversalBFS.java
// Time Complexity : O(N)
// Space Complexity : O(w) maximum width of tree in worst complexity which balance binary tree it is O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class LevelOrderTraversalBFS {
    List<List<Integer>> result;
    public List<List<Integer>> levelOrder(TreeNode root) {
        result = new ArrayList<>();
        if(root == null){
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while(!queue.isEmpty()){

            List<Integer> list = new ArrayList<>();
            int size = queue.size();
            for(int i = 0; i < size; i++){
                TreeNode current = queue.poll();

                if(current.left != null){
                    queue.add(current.left);
                }
                if(current.right != null){
                    queue.add(current.right);
                }
                list.add(current.val);
            }
            result.add(list);
        }

        return result;
    }
}

File: level_order_traversal_bfs.py
# Time Complexity : O(N) where N is number of nodes in the birnary tree
# Space Complexity : O(N) where N is number of nodes in the birnary tree
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque 
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None: return []
        
        q = deque()
        result = []
        q.append(root)
                
        while len(q) != 0:
            size = len(q)
            i = 0
            temp = []
            while i < size:
                i += 1
                node = q.pop()
                temp.append(node.val)
                if node.left is not None:
                    q.appendleft(node.left)
                if node.right is not None:
                    q.appendleft(node.right)
            
            result.append(temp)
            
        return result
                    
                